hal/src/rtl872x/usart_hal.cpp

```diff
@@ -327,27 +339,30 @@ class Usart {
         // Must be called under RX lock!
         if (receiving_ && !useInterrupt()) {
             auto uartInstance = uartTable_[index_].UARTx;
-            size_t transferredToDmaFromUart = uartInstance->RX_BYTE_CNT & RUART_RX_READ_BYTE_CNTER;
             size_t dmaAvailableInBuffer = GDMA_GetDstAddr(rxDmaInitStruct_.GDMA_Index, rxDmaInitStruct_.GDMA_ChNum) - rxDmaInitStruct_.GDMA_DstAddr;
             size_t blockSize = rxDmaInitStruct_.GDMA_BlockSize;
             size_t alreadyCommitted = blockSize - rxBuffer_.acquirePending();
+            size_t transferredToDmaFromUart = uartInstance->RX_BYTE_CNT & RUART_RX_READ_BYTE_CNTER;
+            if (hal_interrupt_is_isr()) {
+                // This method is called from DMA RX ISR
+                // Clear UART RX FIFO read counter ASAP, because UART might be still receiving data 
+                uartInstance->RX_BYTE_CNT |= RUART_RX_BYTE_CNTER_CLEAR;
```

Comment:
This should not be necessary to perform here. DMA will not be reading data out of UART FIFO if the transaction just finished. The counter will stay the same. This is not data put into UART FIFO, this is data transmitted from UART FIFO into DMA FIFO.

Reply:
It is possible that other interrupts with higher priority may interrupt the executtion of the UART RX DMA ISR, while the UART is receiving data. If we clear the counter later in the DMA ISR, we might lose some bytes, especially with higher baudrate.

Reply:
We won't. We are resetting RX_BYTE_CNT only at the end of the DMA transaction. Once DMA transaction finishes RX_BYTE_CNT will no longer increase because DMA no longer transfers data out of UART FIFO into its internal FIFO.

Reply:
I  see 👍 . Accordding to the `REG_RX_BYTE_CNT`'s description:
>Counting the byte number of data reading from Rx FIFO

Reply:
Yes, it's an actual number of bytes already _transferred from UART FIFO into DMA FIFO_, so already in DMA FIFO.